File: tools/sit/build-publish-payload.js
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
import { mkdir, stat, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import {
  appendGithubOutput,
  isMain,
  parseOptionArgs,
  readJsonFile,
  requireOption,
  resolvePathUnder,
  writeJsonFile,
} from './common.js';

export const COMMENT_MARKER = '<!-- sit-pr-fps-summary -->';

export function formatRate(value) {
  if (value === null || value === undefined) {
    return 'n/a';
  }
  return `${Number(value).toFixed(1)}%`;
}

export function formatCount(value) {
  if (value === null || value === undefined) {
    return 'n/a';
  }
  return String(value);
}

export function sanitizeClusterName(value) {
  return String(value || '(unnamed cluster)')
    .replaceAll('|', '\\|')
    .replaceAll('\n', ' ')
    .trim();
}

export function resolveWorkflowOutcome(hasRules, sitResult, diffsitResult, fpsResult) {
  if (!hasRules) {
    return {
      state: 'success',
      description: 'No supported SonarJS rule changes detected',
      workflowStatus: 'Skipped (no changed supported rules)',
    };
  }
  if (sitResult !== 'success') {
    return { state: 'failure', description: 'SIT export failed', workflowStatus: 'Failed' };
  }
  if (diffsitResult !== 'success') {
    return { state: 'failure', description: 'DiffSIT failed', workflowStatus: 'Failed' };
  }
  if (fpsResult !== 'success') {
    return {
      state: 'failure',
      description: 'FPS failed for one or more rules',
      workflowStatus: 'Failed',
    };
  }
  return {
    state: 'success',
    description: 'SIT/FPS/DiffSIT completed successfully',
    workflowStatus: 'Success',
  };
}

export function buildCommentLines(context, workflowStatus) {
  const lines = [
    COMMENT_MARKER,
    '',
    '## SIT/FPS PR Automation',
    '',
    `- Status: **${workflowStatus}**`,
    `- Workflow run: ${context.runUrl}`,
    `- Plugin version: \`${context.pluginVersion || 'n/a'}\``,
    `- Languages: \`${context.languages.length > 0 ? context.languages.join(', ') : 'none'}\``,
    `- Changed rules: \`${context.fullRuleKeys.length > 0 ? context.fullRuleKeys.join(', ') : 'none'}\``,
  ];

  if (context.hasRules) {
    lines.push(
      `- SIT artifacts: ${[
        renderArtifactLink('sit-export-target', context.artifactLinks),
        renderArtifactLink('sit-export-baseline', context.artifactLinks),
      ].join(', ')}`,
      `- FPS artifacts: ${[
        renderArtifactLink('fps-reports', context.artifactLinks),
        renderArtifactLink('fps-run-summary', context.artifactLinks),
      ].join(', ')}`,
      `- DiffSIT artifacts: ${[
        renderArtifactLink('diffsit-reports', context.artifactLinks),
        renderArtifactLink('diffsit-run-summary', context.artifactLinks),
      ].join(', ')}`,
    );
  } else {
    lines.push('- SIT artifacts: `not produced`');
    lines.push('- FPS artifacts: `not produced`');
    lines.push('- DiffSIT artifacts: `not produced`');
  }

  if (context.fpsSummary?.results?.length > 0) {
    lines.push(...renderRuleSections(context.fpsSummary.results));
  }
  if (context.diffsitSummary?.overall) {
    lines.push(...renderDiffsitSummary(context.diffsitSummary));
  }
  return lines;
}

function renderRuleSections(results) {
  const lines = ['', '### FPS per rule', ''];
  for (const result of [...results].sort((a, b) =>
    String(a.rule_key ?? '').localeCompare(String(b.rule_key ?? '')),
  )) {
    const ruleKey = result.rule_key ?? 'unknown';
    const status = result.status ?? 'unknown';
    const issuesAnalyzed = formatCount(result.issues_analyzed);
    const fpRate = formatRate(result.false_positive_rate);
    lines.push(
      `#### \`${ruleKey}\` - \`${status}\` - ${issuesAnalyzed} issues - FP rate ${fpRate}`,
      '',
    );

    const clusters = result.clusters ?? [];
    if (clusters.length === 0) {
      lines.push('No cluster data (FPS failed or report missing).', '');
      continue;
    }

    lines.push('| Cluster | Issues | FP |', '|:--------|-------:|---:|');
    let totalIssues = 0;
    for (const cluster of [...clusters].sort(
      (a, b) => Number(b.cluster_fp_rate ?? 0) - Number(a.cluster_fp_rate ?? 0),
    )) {
      const issueCount = cluster.issue_count;
      if (issueCount !== null && issueCount !== undefined) {
        totalIssues += Number(issueCount);
      }
      lines.push(
        `| ${sanitizeClusterName(cluster.cluster_name)} | ${formatCount(issueCount)} | ${formatRate(
          cluster.cluster_fp_rate,
        )} |`,
      );
    }
    lines.push(`| **Total** | **${totalIssues}** | **${fpRate}** |`, '');
  }
  return lines;
}

function renderDiffsitSummary(summary) {
  const overall = summary.overall ?? {};
  return [
    '',
    '### DiffSIT summary',
    '',
    '| Projects | Base issues | Target issues | New | Changed | Message changes | Secondary changes | Removed | Unchanged |',
    '|---------:|------------:|--------------:|----:|--------:|----------------:|------------------:|--------:|----------:|',
    `| ${formatCount(overall.projects)} | ${formatCount(overall.base_count)} | ${formatCount(
      overall.target_count,
    )} | ${formatCount(overall.new)} | ${formatCount(overall.changed)} | ${formatCount(
      overall.message_changes,
    )} | ${formatCount(overall.secondary_changes)} | ${formatCount(overall.removed)} | ${formatCount(
      overall.unchanged,
    )} |`,
    '',
  ];
}

export function buildRuleStatuses(results, targetUrl) {
  const statuses = [];
  for (const result of results) {
    const ruleKey = result.rule_key;
    if (!ruleKey) {
      continue;
    }
    const state = result.status === 'success' ? 'success' : 'failure';
    const metrics = [];
    if (result.issues_analyzed !== null && result.issues_analyzed !== undefined) {
      metrics.push(`issues=${result.issues_analyzed}`);
    }
    if (result.false_positive_rate !== null && result.false_positive_rate !== undefined) {
      metrics.push(`fp=${Number(result.false_positive_rate).toFixed(1)}%`);
    }
    if (result.cluster_count !== null && result.cluster_count !== undefined) {
      metrics.push(`clusters=${result.cluster_count}`);
    }
    const statusWord = state === 'success' ? 'success' : 'failed';
    const description = truncateDescription(
      `FPS ${statusWord}${metrics.length > 0 ? ` (${metrics.join(', ')})` : ''}`,
    );
    statuses.push({
      context: `sit-fps/${ruleKey}`,
      state,
      description,
      target_url: targetUrl,
    });
  }
  return statuses;
}

export async function loadArtifactLinks(artifactsJsonPath, repository, runId) {
  if (!(await exists(artifactsJsonPath))) {
    return {};
  }
  let payload;
  try {
    payload = await readJsonFile(artifactsJsonPath, 'artifacts JSON');
  } catch {
    return {};
  }

  const links = {};
  for (const artifact of payload.artifacts ?? []) {
    if (artifact.id === undefined || !artifact.name) {
      continue;
    }
    links[artifact.name] =
      `https://github.com/${repository}/actions/runs/${runId}/artifacts/${artifact.id}`;
  }
  return links;
}

function renderArtifactLink(artifactName, artifactLinks) {
  const artifactLink = artifactLinks[artifactName];
  if (!artifactLink) {
    return `\`${artifactName}\` (artifact not found)`;
  }
  return `[\`${artifactName}\`](${artifactLink})`;
}

export async function buildPublishPayload(args, appendOutput = appendGithubOutput) {
  const rules = loadRules(args.rulesJson);
  const fpsSummary = await loadOptionalJson(args.fpsSummaryPath);
  const diffsitSummary = await loadOptionalJson(args.diffsitSummaryPath);
  const artifactLinks = await loadArtifactLinks(
    args.artifactsJsonPath,
    args.repository,
    args.runId,
  );
  const context = {
    rules,
    hasRules: rules.length > 0,
    languages: deriveLanguages(rules),
    fullRuleKeys: deriveFullRuleKeys(rules),
    sitResult: args.sitResult || 'skipped',
    diffsitResult: args.diffsitResult || 'skipped',
    fpsResult: args.fpsResult || 'skipped',
    pluginVersion: args.pluginVersion || 'n/a',
    runUrl: args.runUrl,
    fpsSummary,
    diffsitSummary,
    artifactLinks,
  };
  const outcome = resolveWorkflowOutcome(
    context.hasRules,
    context.sitResult,
    context.diffsitResult,
    context.fpsResult,
  );
  await writeComment(args.commentPath, buildCommentLines(context, outcome.workflowStatus));
  await writeJsonFile(
    args.ruleStatusesPath,
    fpsSummary?.results ? buildRuleStatuses(fpsSummary.results, args.runUrl) : [],
  );
  await appendOutput('state', outcome.state);
  await appendOutput('description', outcome.description);
  return { outcome, context };
}

function loadRules(rulesJson) {
  const rawRules = JSON.parse(rulesJson || '[]');
  if (!Array.isArray(rawRules)) {
    throw new Error('rules JSON must be an array');
  }
  return rawRules.map(rule => ({
    repository: String(rule.repository).toLowerCase(),
    language: String(rule.language).toLowerCase(),
    ruleKey: String(rule.ruleKey ?? rule.rule_key).toUpperCase(),
  }));
}

function deriveLanguages(rules) {
  return [...new Set(rules.map(rule => rule.language))].sort();
}

function deriveFullRuleKeys(rules) {
  return [...new Set(rules.map(rule => `${rule.repository}:${rule.ruleKey}`))].sort();
}

async function writeComment(path, lines) {
  await writeJsonOrText(path, `${lines.join('\n').trim()}\n`);
}

async function writeJsonOrText(path, text) {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, text, 'utf8');
}

async function loadOptionalJson(path) {
  if (!(await exists(path))) {
    return null;
  }
  return readJsonFile(path);
}

async function exists(path) {
  try {
    await stat(path);
    return true;
  } catch (error) {
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}

function truncateDescription(description, maxLen = 140) {
  if (description.length <= maxLen) {
    return description;
  }
  return `${description.slice(0, maxLen - 3)}...`;
}

async function main() {
  const options = parseOptionArgs(process.argv);
  const cwd = process.cwd();
  await buildPublishPayload({
    rulesJson: requireOption(options, '--rules-json'),
    sitResult: requireOption(options, '--sit-result'),
    diffsitResult: requireOption(options, '--diffsit-result'),
    fpsResult: requireOption(options, '--fps-result'),
    pluginVersion: options.get('--plugin-version') ?? '',
    runUrl: requireOption(options, '--run-url'),
    repository: requireOption(options, '--repository'),
    runId: requireOption(options, '--run-id'),
    fpsSummaryPath: resolvePathUnder(
      cwd,
      requireOption(options, '--fps-summary-path'),
      '--fps-summary-path',
    ),
    diffsitSummaryPath: resolvePathUnder(
      cwd,
      requireOption(options, '--diffsit-summary-path'),
      '--diffsit-summary-path',
    ),
    artifactsJsonPath: resolvePathUnder(
      cwd,
      requireOption(options, '--artifacts-json-path'),
      '--artifacts-json-path',
    ),
    commentPath: resolvePathUnder(cwd, requireOption(options, '--comment-path'), '--comment-path'),
    ruleStatusesPath: resolvePathUnder(
      cwd,
      requireOption(options, '--rule-statuses-path'),
      '--rule-statuses-path',
    ),
  });
}

if (isMain(import.meta.url)) {
  main().catch(error => {
    console.error(error instanceof Error ? error.message : error);
    process.exit(1);
  });
}

File: dev-tools/scripts/sit_pr_fps/build_publish_payload.py
"""Build publish artifacts for SIT/FPS PR automation.

Generates the commit-status outcome, per-rule status payload, and markdown
comment body consumed by the workflow publish step.
"""

import argparse
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from common import append_output, resolve_path_under

COMMENT_MARKER = "<!-- sit-pr-fps-summary -->"


def path_under_cwd(label: str):
    def parse_path(value: str) -> Path:
        try:
            return resolve_path_under(Path.cwd(), value, label)
        except ValueError as err:
            raise argparse.ArgumentTypeError(str(err)) from err

    return parse_path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Build commit-status and PR comment payload for SIT/FPS workflow")
    parser.add_argument("--rules-json", required=True)
    parser.add_argument("--sit-result", required=True)
    parser.add_argument("--diffsit-result", required=True)
    parser.add_argument("--fps-result", required=True)
    parser.add_argument(
        "--plugin-version",
        default="",
        help="Analyzer plugin version to include in the published comment.",
    )
    parser.add_argument(
        "--rule-key-format",
        default="{language}dre:{rule_id}",
        help=(
            "Python format string used to build full rule keys from `language` "
            "and `rule_id` fields of rules_json (default keeps sonar-skunk's "
            "DRE convention)."
        ),
    )
    parser.add_argument(
        "--sit-artifact-name-template",
        default="sit-export-bundles-{language}",
        help=(
            "Template for per-language SIT artifact names used when rendering "
            "links in the PR comment (default keeps sonar-skunk's per-language "
            "fan-out convention)."
        ),
    )
    parser.add_argument("--run-url", required=True)
    parser.add_argument("--repository", required=True)
    parser.add_argument("--run-id", required=True)
    parser.add_argument("--fps-summary-path", required=True, type=path_under_cwd("--fps-summary-path"))
    parser.add_argument("--diffsit-summary-path", required=True, type=path_under_cwd("--diffsit-summary-path"))
    parser.add_argument("--artifacts-json-path", required=True, type=path_under_cwd("--artifacts-json-path"))
    parser.add_argument("--comment-path", required=True, type=path_under_cwd("--comment-path"))
    parser.add_argument("--rule-statuses-path", required=True, type=path_under_cwd("--rule-statuses-path"))
    return parser.parse_args()


def load_json(value: str, fallback: Any) -> Any:
    if not value:
        return fallback
    return json.loads(value)


def derive_languages(rules: list[dict[str, str]]) -> list[str]:
    return sorted({rule["language"] for rule in rules})


def derive_full_rule_keys(rules: list[dict[str, str]], rule_key_format: str) -> list[str]:
    return [rule_key_format.format(language=rule["language"], rule_id=rule["rule_id"]) for rule in rules]


@dataclass(frozen=True)
class BuildContext:
    rules: list[dict[str, str]]
    has_rules: bool
    languages: list[str]
    full_rule_keys: list[str]
    sit_result: str
    diffsit_result: str
    fps_result: str
    plugin_version: str
    run_url: str
    repository: str
    run_id: str
    fps_summary: dict[str, Any] | None
    diffsit_summary: dict[str, Any] | None
    artifact_links: dict[str, str]
    sit_artifact_name_template: str


@dataclass(frozen=True)
class WorkflowOutcome:
    state: str
    description: str
    workflow_status: str


def format_rate(value: Any) -> str:
    if value is None:
        return "n/a"
    return f"{float(value):.1f}%"


def format_count(value: Any) -> str:
    if value is None:
        return "n/a"
    return str(value)


def sanitize_cluster_name(value: Any) -> str:
    cluster_name = str(value or "(unnamed cluster)")
    return cluster_name.replace("|", "\\|").replace("\n", " ").strip()


def render_rule_sections(results: list[dict[str, Any]]) -> list[str]:
    lines = ["", "### FPS per rule", ""]
    sorted_results = sorted(results, key=lambda result: str(result.get("rule_key", "")))

    for result in sorted_results:
        rule_key = result.get("rule_key", "unknown")
        status = result.get("status", "unknown")
        issues_analyzed = format_count(result.get("issues_analyzed"))
        fp_rate = format_rate(result.get("false_positive_rate"))

        lines.append(f"#### `{rule_key}` — `{status}` — {issues_analyzed} issues — FP rate {fp_rate}")
        lines.append("")

        clusters = result.get("clusters") or []
        if not clusters:
            lines.append("No cluster data (FPS failed or report missing).")
            lines.append("")
            continue

        sorted_clusters = sorted(
            clusters,
            key=lambda cluster: cluster.get("cluster_fp_rate") or 0,
            reverse=True,
        )

        lines.extend(["| Cluster | Issues | FP |", "|:--------|-------:|---:|"])

        total_issues = 0
        for cluster in sorted_clusters:
            name = sanitize_cluster_name(cluster.get("cluster_name"))
            issue_count = cluster.get("issue_count")
            if issue_count is not None:
                total_issues += issue_count
            lines.append(f"| {name} | {format_count(issue_count)} | {format_rate(cluster.get('cluster_fp_rate'))} |")

        lines.append(f"| **Total** | **{total_issues}** | **{fp_rate}** |")
        lines.append("")

    return lines


def render_diffsit_summary(summary: dict[str, Any]) -> list[str]:
    overall = summary.get("overall") or {}
    return [
        "",
        "### DiffSIT summary",
        "",
        (
            "| Projects | Base issues | Target issues | New | Changed | Message changes "
            "| Secondary changes | Removed | Unchanged |"
        ),
        "|---------:|------------:|--------------:|----:|--------:|----------------:|------------------:|--------:|----------:|",
        (
            f"| {format_count(overall.get('projects'))} "
            f"| {format_count(overall.get('base_count'))} "
            f"| {format_count(overall.get('target_count'))} "
            f"| {format_count(overall.get('new'))} "
            f"| {format_count(overall.get('changed'))} "
            f"| {format_count(overall.get('message_changes'))} "
            f"| {format_count(overall.get('secondary_changes'))} "
            f"| {format_count(overall.get('removed'))} "
            f"| {format_count(overall.get('unchanged'))} |"
        ),
        "",
    ]


def truncate_description(description: str, max_len: int = 140) -> str:
    if len(description) <= max_len:
        return description
    return description[: max_len - 3] + "..."


def load_artifact_links(artifacts_json_path: Path, repository: str, run_id: str) -> dict[str, str]:
    artifacts_json_path = resolve_path_under(Path.cwd(), artifacts_json_path, "--artifacts-json-path")
    if not artifacts_json_path.exists():
        return {}

    try:
        payload = json.loads(artifacts_json_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return {}

    links = {}
    for artifact in payload.get("artifacts", []):
        artifact_id = artifact.get("id")
        artifact_name = artifact.get("name")
        if artifact_id is None or not artifact_name:
            continue
        links[artifact_name] = f"https://github.com/{repository}/actions/runs/{run_id}/artifacts/{artifact_id}"
    return links


def render_artifact_link(artifact_name: str, artifact_links: dict[str, str]) -> str:
    artifact_link = artifact_links.get(artifact_name)
    if not artifact_link:
        return f"`{artifact_name}` (artifact not found)"
    return f"[`{artifact_name}`]({artifact_link})"


def build_rule_statuses(results: list[dict[str, Any]], target_url: str) -> list[dict[str, str]]:
    statuses = []
    for result in results:
        rule_key = result.get("rule_key")
        if not rule_key:
            continue

        state = "success" if result.get("status") == "success" else "failure"
        metrics = []
        issues = result.get("issues_analyzed")
        fp_rate = result.get("false_positive_rate")
        clusters = result.get("cluster_count")
        if issues is not None:
            metrics.append(f"issues={issues}")
        if fp_rate is not None:
            metrics.append(f"fp={float(fp_rate):.1f}%")
        if clusters is not None:
            metrics.append(f"clusters={clusters}")

        status_word = "success" if state == "success" else "failed"
        description = f"FPS {status_word}"
        if metrics:
            description += f" ({', '.join(metrics)})"

        statuses.append(
            {
                "context": f"sit-fps/{rule_key}",
                "state": state,
                "description": truncate_description(description),
                "target_url": target_url,
            }
        )

    return statuses


def load_fps_summary(fps_summary_path: Path) -> dict[str, Any] | None:
    fps_summary_path = resolve_path_under(Path.cwd(), fps_summary_path, "--fps-summary-path")
    if not fps_summary_path.exists():
        return None
    return json.loads(fps_summary_path.read_text(encoding="utf-8"))


def load_diffsit_summary(diffsit_summary_path: Path) -> dict[str, Any] | None:
    diffsit_summary_path = resolve_path_under(Path.cwd(), diffsit_summary_path, "--diffsit-summary-path")
    if not diffsit_summary_path.exists():
        return None
    return json.loads(diffsit_summary_path.read_text(encoding="utf-8"))


def load_context(args: argparse.Namespace) -> BuildContext:
    rules = load_json(args.rules_json, [])
    return BuildContext(
        rules=rules,
        has_rules=len(rules) > 0,
        languages=derive_languages(rules),
        full_rule_keys=derive_full_rule_keys(rules, args.rule_key_format),
        sit_result=args.sit_result or "skipped",
        diffsit_result=args.diffsit_result or "skipped",
        fps_result=args.fps_result or "skipped",
        plugin_version=args.plugin_version or "n/a",
        run_url=args.run_url,
        repository=args.repository,
        run_id=args.run_id,
        fps_summary=load_fps_summary(args.fps_summary_path),
        diffsit_summary=load_diffsit_summary(args.diffsit_summary_path),
        artifact_links=load_artifact_links(args.artifacts_json_path, args.repository, args.run_id),
        sit_artifact_name_template=args.sit_artifact_name_template,
    )


def resolve_workflow_outcome(has_rules: bool, sit_result: str, diffsit_result: str, fps_result: str) -> WorkflowOutcome:
    if not has_rules:
        return WorkflowOutcome(
            state="success",
            description="No rule changes detected",
            workflow_status="Skipped (no changed rules)",
        )
    if sit_result != "success":
        return WorkflowOutcome(
            state="failure",
            description="SIT export failed",
            workflow_status="Failed",
        )
    if diffsit_result != "success":
        return WorkflowOutcome(
            state="failure",
            description="DiffSIT failed",
            workflow_status="Failed",
        )
    if fps_result != "success":
        return WorkflowOutcome(
            state="failure",
            description="FPS failed for one or more rules",
            workflow_status="Failed",
        )
    return WorkflowOutcome(
        state="success",
        description="SIT/FPS/DiffSIT completed successfully",
        workflow_status="Success",
    )


def build_comment_lines(context: BuildContext, workflow_status: str) -> list[str]:
    lines = [
        COMMENT_MARKER,
        "",
        "## SIT/FPS PR Automation",
        "",
        f"- Status: **{workflow_status}**",
        f"- Workflow run: {context.run_url}",
        f"- Plugin version: `{context.plugin_version}`",
        f"- Languages: `{', '.join(context.languages) if context.languages else 'none'}`",
        f"- Changed rules: `{', '.join(context.full_rule_keys) if context.full_rule_keys else 'none'}`",
    ]

    if context.languages:
        sit_artifacts = [
            render_artifact_link(
                context.sit_artifact_name_template.format(language=language),
                context.artifact_links,
            )
            for language in context.languages
        ]
        lines.append(f"- SIT artifacts: {', '.join(sit_artifacts)}")
    else:
        lines.append("- SIT artifacts: `none`")

    if context.has_rules:
        fps_artifacts = [
            render_artifact_link("fps-reports", context.artifact_links),
            render_artifact_link("fps-run-summary", context.artifact_links),
        ]
        lines.append(f"- FPS artifacts: {', '.join(fps_artifacts)}")
        diffsit_artifacts = [
            render_artifact_link("diffsit-reports", context.artifact_links),
            render_artifact_link("diffsit-run-summary", context.artifact_links),
        ]
        lines.append(f"- DiffSIT artifacts: {', '.join(diffsit_artifacts)}")
    else:
        lines.append("- FPS artifacts: `not produced`")
        lines.append("- DiffSIT artifacts: `not produced`")

    if context.fps_summary and context.fps_summary.get("results"):
        lines.extend(render_rule_sections(context.fps_summary["results"]))
    if context.diffsit_summary and context.diffsit_summary.get("overall"):
        lines.extend(render_diffsit_summary(context.diffsit_summary))
    return lines


def write_file(path: Path, content: str) -> None:
    path = resolve_path_under(Path.cwd(), path, "output path")
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def write_comment(path: Path, lines: list[str]) -> None:
    write_file(path, "\n".join(lines).strip() + "\n")


def write_rule_statuses(path: Path, rule_statuses: list[dict[str, str]]) -> None:
    write_file(path, json.dumps(rule_statuses, indent=2))


def publish_workflow_outputs(outcome: WorkflowOutcome) -> None:
    append_output("state", outcome.state)
    append_output("description", outcome.description)


def main() -> int:
    args = parse_args()
    context = load_context(args)
    outcome = resolve_workflow_outcome(
        has_rules=context.has_rules,
        sit_result=context.sit_result,
        diffsit_result=context.diffsit_result,
        fps_result=context.fps_result,
    )

    comment_lines = build_comment_lines(context=context, workflow_status=outcome.workflow_status)
    write_comment(path=args.comment_path, lines=comment_lines)

    rule_statuses: list[dict[str, str]] = []
    if context.fps_summary and context.fps_summary.get("results"):
        rule_statuses = build_rule_statuses(context.fps_summary["results"], context.run_url)
    write_rule_statuses(path=args.rule_statuses_path, rule_statuses=rule_statuses)

    publish_workflow_outputs(outcome)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
